File: 19/part2.py
import regex
rules, msgs = open("input.txt").read().split("\n\n")
rule_dict = dict( tuple(r.split(": ")) for r in rules.split('\n') )

def build(num):
    if num == "8":
        return "(?:" + build("42") + ")+"
    elif num == "11":
        return "(?P<one>" + build("42") + build("31") + "|" + build("42") + "(?P>one)" + build("31") + ")"
    group = "(?:"
    for t in rule_dict[num].split():
        if t[0] == '"':
            group = group + t[1]
        elif t[0] in ["0","1","2","3","4","5","6","7","8","9"]:
            group = group + build(t)
        elif t[0] == '|':
            group = group + '|'
    return group + ')'

zero_re = regex.compile("^" + build("0") + "$")

s = 0
for msg in msgs.split('\n'):
    m = zero_re.match(msg)
    if not (m is None):
        s += 1

print(s)

File: 19/part2.jl
rules, msgs = split(read("input.txt",String),"\n\n")
rule_dict = Dict( Pair(split(r,": ")...) for r in split(rules,'\n') )

function build(num)
   if num == "8"
      return "(?:" * build("42") * ")+"
   elseif num == "11"
      return "(?P<eleven>" * build("42") * build("31") * "|" * build("42") * "(?P>eleven)" * build("31") * ")"
   end

   group = String("(?:")
   for t in split(rule_dict[num])
      if t[1] == '"'
         group = group * t[2]
      elseif t[1] in '0':'9'
         group = group * build(t)
      elseif t[1] == '|'
         group = group * '|'
      end
   end
   group * ')'
end

println( sum( occursin(Regex('^' * build("0") * '$'), msg) for msg in split(msgs,'\n') ) )
